text_generation/causal_lm/cpp/continuous_batching/library/src/sampler.hpp

```diff
@@ -401,34 +405,55 @@ SamplerOutput Sampler::sample(std::vector<SequenceGroup::Ptr> & sequence_groups,
+                        sampler_output.m_dropped_sequences.push_back(running_sequence->get_id());
+                    }
+                };
+                for (size_t running_sequence_id = 0; running_sequence_id < num_running_sequences; ++running_sequence_id) {
+                    auto logit_vector = _get_logit_vector(sequence_group_logits, running_sequence_id);
 
-                sequence_group->register_generated_token_id(sampled_token_id);
-
-                // in case of greedy search we always have a single parent sequence to sample from
-                running_sequences[0]->append_token(sampled_token_id, sequence_group_logits.data<const float>()[sampled_token_id]);
+                    if (sampling_params.repetition_penalty != 1.0f) {
+                        auto repetition_penalty_transform = RepetitionPenaltyTransform(sampling_params.repetition_penalty);
+                        logit_vector = repetition_penalty_transform.apply(logit_vector, sequence_group->get_unique_generated_ids());
+                    }
 
-                if (sampling_params.max_new_tokens == running_sequences[0]->get_generated_len() ||
-                    sampled_token_id == sampling_params.eos_token_id && !sampling_params.ignore_eos) {
-                    // stop sequence by max_new_tokens or EOS token
-                    running_sequences[0]->set_status(SequenceStatus::FINISHED);
-                    // drop sequence from scheduler
-                    sampler_output.m_dropped_sequences.push_back(running_sequences[0]->get_id());
+                    LogitWithIdx sampled_token_id;
+                    if (sampling_params.is_greedy_sampling()) {
+                        sampled_token_id = _greedy_sample(logit_vector);
+                    } else {
+                        // is_multinomial()
+                        const bool is_generate_n_tokens = num_running_sequences == 1  && sequence_group->get_finished_sequences().empty();
+                        const size_t num_tokens_per_sequence = is_generate_n_tokens ? sampling_params.num_return_sequences : 1;
+                        auto sampled_token_ids = _multinomial_sample(logit_vector, sampling_params.temperature, sampling_params.top_p,
+                                                                     sampling_params.top_k, num_tokens_per_sequence);
+                        sampled_token_id = sampled_token_ids[0];
+
+                        if (is_generate_n_tokens) {
+                            auto sequence_to_fork = running_sequences[0];
+                            std::list<uint64_t> forked_seq_ids;
+                            for (size_t i = num_running_sequences; i < num_tokens_per_sequence; ++i) {
+                                const auto forked_sequence = sequence_group->fork_sequence(sequence_to_fork);
+                                forked_seq_ids.push_back(forked_sequence->get_id());
+                                running_sequences.push_back(forked_sequence);
```

Comment:
looks like we don't need to push to `running_sequences`, because sequence group is updated by `sequence_group->fork_sequence`